internal/schema/usage_data.go

```diff
@@ -51,6 +52,45 @@ func (u *UsageData) Get(key string) gjson.Result {
 	return u.Attributes[key]
 }
 
+// GetMap returns a map of gjson.Result. GetMap expects key to be
+// the name of a usage file property that contains sub properties. e.g.
+//
+// 		given a usage file with the following property:
+//
+//			prop_1:
+//				sub_prop_1: 1000
+//				sub_prop_2: 3000
+//
+// 		GetMap with the "prop_1" as key will return a map:
+//
+//			map[string]gjson.Result{
+//				"sub_prop_1": ...,
+//				"sub_prop_2": ...,
+//		   }
+//
+// GetMap only support 1 level of nesting. Further levels must be handled another way.
+// If key cannot be found in UsageData GetMap returns nil.
+func (u *UsageData) GetMap(key string) map[string]gjson.Result {
```

Comment:
Can this be done with `u.Get(key).Map()`?

Reply:
Sadly not as at the point of fetching this data (in the Resource) we have flattened this data. See these lines here: https://github.com/infracost/infracost/blob/285d9a072a4bec6ccd2cd6a1904dff2900c8f7b8/internal/schema/usage_data.go#L137-L168 .

This means that instead the expected result from gjson parsing of:

```go
UsageData{
    Attributes: map[string]gjson.Result{
        "monthly_outbound_from_region_to_dx_connection_location": gjson.Result{
            Type: "JSON",
            Raw: `{"ap_east_1": 3000}`
        }
    }
}
```

which support `Result.Map` - see here: https://github.com/tidwall/gjson/blob/590010fdac311cc8990ef5c97448d4fec8f29944/gjson.go#L289-L295

We get the following:

```go
UsageData{
    Attributes: map[string]gjson.Result{
        "monthly_outbound_from_region_to_dx_connection_location.ap_east_1": gjson.Result{
            Type: "JSON",
            Raw: `3000`
            Num: 3000
        }
    }
}
```

So `GetMap` supports unflattening from 1 level of nesting. As mentioned in the PR description I think it would be best in the future to try and remove flattening from the start. However I didn't want to do it in this PR as it basically touches every resource. Maybe we should put it into an issue to track?

Reply:
Would it work if we invert things and look for the specific keys, so in `NewDXConnection` do something like:
```go
if u != nil { 
    // sort the region keys so that we get a consistent output in the cli
    keys := make([]string, 0, len(regions))
    ...
    
    for _, key := range keys {
         if u.GetMap("monthly_outbound_from_region_to_dx_connection_location." + key ) != nil {
             // create the cost component
         }
    }
```
 I'm a little wary of adding GetMap right now since it seems like there's a lot of stuff inflight wrt usage params.  Also at some point we're going to port this to the 'provder agnostic' style which has a defined "UsageSchema" and I think it may be more straightforward to do that if we treat each value individually instead of as a map (take a look at `LambdaFunctionUsageSchema` in `internal/resources/aws/lambda_function.go` as an example)

Reply:
@tim775 sure thing can scrap the `GetMap` function and just use base `Get` with the key lookup as suggested. I'll change now. Just checking you meant to use: `u.Get("monthly_outbound_from_region_to_dx_connection_location." + key )` in the above example instead of `GetMap`

Reply:
Oops, yes I meant `u.Get("monthly_outbound_from_region_to_dx_connection_location." + key )` (I edited my comment)

Reply:
Perf - I'll use the defined `regionMapping` to define a list of acceptable `keys` and then attempt to pull each using `Get` 👍